src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingAssignmentOperator.cs

```diff
@@ -14,35 +16,143 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
+                                               kind: SynthesizedLocalKind.Spill);
+
+                temps.Add(tmp.LocalSymbol);
+
+                // tmp = loweredRight;
+                var tmpAssignment =
+                    MakeAssignmentOperator(node.Syntax, tmp, loweredRight, node.Type, used: true, isChecked: false, isCompoundAssignment: false);
+
+                // transformedLhs = tmp;
+                var transformedLhsAssignment =
+                    MakeAssignmentOperator(
+                        node.Syntax,
+                        transformedLHS,
+                        MakeConversionNode(tmp, transformedLHS.Type, @checked: false),
+                        node.LeftOperand.Type,
+                        used: true,
+                        isChecked: false,
+                        isCompoundAssignment: false); ;
+
+                // !lhsRead.HasValue
+                var lhsReadHasValue = BoundCall.Synthesized(leftOperand.Syntax, lhsRead, hasValue);
+                var lhsReadHasValueNegation =
+                    MakeUnaryOperator(UnaryOperatorKind.BoolLogicalNegation, leftOperand.Syntax, method: null, lhsReadHasValue, lhsReadHasValue.Type);
+
+                // { tmp = b; transformedLhs = tmp; }
+                var consequenceBuilder = ArrayBuilder<BoundStatement>.GetInstance(2);
+                consequenceBuilder.Add(new BoundExpressionStatement(node.Syntax, tmpAssignment) { WasCompilerGenerated = true });
+                consequenceBuilder.Add(new BoundExpressionStatement(node.Syntax, transformedLhsAssignment) { WasCompilerGenerated = true });
+                var consequence = new BoundBlock(node.Syntax, locals: ImmutableArray<LocalSymbol>.Empty, consequenceBuilder.ToImmutableAndFree())
+                {
+                    WasCompilerGenerated = true
+                };
+
+                // if (!lhsRead.HasValue) { tmp = b; transformedLhs = tmp; }
+                var ifHasValueStatement =
+                    RewriteIfStatement(node.Syntax, lhsReadHasValueNegation, consequence, rewrittenAlternativeOpt: null, hasErrors: false);
+
+                // Assemble final statement list, including any stores that were originally used
+                // for capturing the lhs in a temp to start with.
+                var statementsBuilder = ArrayBuilder<BoundStatement>.GetInstance(2 + stores.Count);
+                foreach (var store in stores)
+                {
+                    statementsBuilder.Add(new BoundExpressionStatement(store.Syntax, store) { WasCompilerGenerated = true });
+                }
+                stores.Free();
+                statementsBuilder.Add(new BoundExpressionStatement(leftOperand.Syntax, getValueOrDefaultStore) { WasCompilerGenerated = true });
+                statementsBuilder.Add(ifHasValueStatement);
+
+                _needsSpilling = true;
+                return _factory.SpillSequence(temps.ToImmutableAndFree(), statementsBuilder.ToImmutableAndFree(), tmp);
```

Comment:
>SpillSequence [](start = 32, length = 13)

Technically, we could do without the spilling, just using a `Sequence`. I don't know if it's worth it though.
```
// no statements, only expressions and side-effects in the following:
// var tmp = lhsRead.GetValueOrDefault();
// !lhsRead.HasValue ? { /*sequence*/ tmp = loweredRight; transformedAssignment = tmp; tmp } : tmp
```

Reply:
I think this version is likely more efficient IL wise, since it doesn't have the multiple `tmp`s at the end of a ternary. Also, it's the version that LDM approved, so I'd prefer to stick with the existing one.

---
In reply to: [293099162](https://github.com/dotnet/roslyn/pull/36329#discussion_r293099162) [](ancestors = 293099162)

Reply:
I don't think LDM approved any specific IL-level lowered form. I recall we only discussed a C#-level lowered form and also guiding principles.
Don't forget that the lowered form here (described at the top of this local function) is not the final lowered form anyways, as the spilling will manipulate it further. Also, spilling makes this harder to reason about (exactly what is spilled? what is the final form?) and required some changes to `TransformCompoundAssignmentLHS` which I'm a bit concerned about.

I'm curious what Neal, Aleksey or Chuck think about this.

---
In reply to: [293103887](https://github.com/dotnet/roslyn/pull/36329#discussion_r293103887) [](ancestors = 293103887,293099162)